ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,437 @@
+    return !!(array[pos / (8 * sizeof(uint32_t))] & (1U << (pos % (8 * sizeof(uint32_t)))));
+}
+
+static inline void CLEAR_SPC_BIT(uint32_t* array, int32_t pos)
+{
+    assert(pos < OMPI_SPC_NUM_COUNTERS);
+    array[pos / (8 * sizeof(uint32_t))] &= ~(1U << (pos % (8 * sizeof(uint32_t))));
+}
+
+/* ##############################################################
+ * ################# Begin MPI_T Functions ######################
+ * ##############################################################
+ */
+static int ompi_spc_notify(mca_base_pvar_t *pvar, mca_base_pvar_event_t event, void *obj_handle, int *count)
+{
+    (void)obj_handle;
+
+    int index;
+
+    if(OPAL_LIKELY(!mpi_t_enabled)) {
+        return MPI_SUCCESS;
+    }
+
+    /* For this event, we need to set count to the number of long long type
+     * values for this counter.  All SPC counters are one long long, so we
+     * always set count to 1.
+     */
+    if(MCA_BASE_PVAR_HANDLE_BIND == event) {
+        *count = 1;
+    }
+    /* For this event, we need to turn on the counter */
+    else if(MCA_BASE_PVAR_HANDLE_START == event) {
+        /* Convert from MPI_T pvar index to SPC index */
+        index = pvar->pvar_index - mpi_t_offset;
+        SET_SPC_BIT(ompi_spc_attached_event, index);
+    }
+    /* For this event, we need to turn off the counter */
+    else if(MCA_BASE_PVAR_HANDLE_STOP == event) {
+        /* Convert from MPI_T pvar index to SPC index */
+        index = pvar->pvar_index - mpi_t_offset;
+        CLEAR_SPC_BIT(ompi_spc_attached_event, index);
+    }
+
+    return MPI_SUCCESS;
+}
+
+/* ##############################################################
+ * ################# Begin SPC Functions ########################
+ * ##############################################################
+ */
+
+/* This function returns the current count of an SPC counter that has been retistered
+ * as an MPI_T pvar.  The MPI_T index is not necessarily the same as the SPC index,
+ * so we need to convert from MPI_T index to SPC index and then set the 'value' argument
+ * to the correct value for this pvar.
+ */
+static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, void *obj_handle)
+{
+    (void) obj_handle;
+
+    long long *counter_value = (long long*)value;
+
+    if(OPAL_LIKELY(!mpi_t_enabled)) {
+        *counter_value = 0;
+        return MPI_SUCCESS;
+    }
+
+    /* Convert from MPI_T pvar index to SPC index */
+    int index = pvar->pvar_index - mpi_t_offset;
+    /* Set the counter value to the current SPC value */
+    *counter_value = (long long)ompi_spc_events[index].value;
+    /* If this is a timer-based counter, convert from cycles to microseconds */
+    if( IS_SPC_BIT_SET(ompi_spc_timer_event, index) ) {
+        *counter_value /= sys_clock_freq_mhz;
+    }
+    /* If this is a high watermark counter, reset it after it has been read */
+    if(index == OMPI_MAX_UNEXPECTED_IN_QUEUE || index == OMPI_MAX_OOS_IN_QUEUE) {
```

Comment:
Per my previous comment:

> It might be better to have some kind of attribute somewhere that indicates that a particular SPC counter needs to be reset after being read (Vs. having a hard-coded list of specific enums here in the code).